### PR TITLE
Self replace setup.sh if it does not match remote

### LINKED ISSUE #293: More congenial setup.sh updates
Could the setup script check that it is the latest version, and if not then perhaps offer to overwrite itself with the latest version? 

To help avoid situations like this:

```
BusyBox v1.36.1 (2024-04-03 00:56:56 UTC) built-in shell (ash)

  _______                     ________        __
 |       |.-----.-----.-----.|  |  |  |.----.|  |_
 |   -   ||  _  |  -__|     ||  |  |  ||   _||   _|
 |_______||   __|_____|__|__||________||__|  |____|
          |__| W I R E L E S S   F R E E D O M
 -----------------------------------------------------
 OpenWrt 23.05.3, r23809-234f1a2efa
 -----------------------------------------------------
root@OpenWrt-1:~# cd cake-autorate/
root@OpenWrt-1:~/cake-autorate# ./setup.sh
At least one instance of cake-autorate appears to be running - exiting
If you want to install a new version, first stop any running instance of cake-autorate
If you are sure that no instance of cake-autorate is running, delete the /var/run/cake-autorate directory
root@OpenWrt-1:~/cake-autorate# service cake-autorate stop
root@OpenWrt-1:~/cake-autorate# ./setup.sh
Installing cake-autorate in /root/cake-autorate...
Previous configuration present - keep it? [Y/n]
mv: can't rename '/tmp/tmp.NJfmff/launcher.sh': No such file or directory
root@OpenWrt-1:~/cake-autorate# ./setup.sh
Installing cake-autorate in /root/cake-autorate...
Previous configuration present - keep it? [Y/n]
mv: can't rename '/tmp/tmp.eihdno/launcher.sh': No such file or directory
root@OpenWrt-1:~/cake-autorate# wget https://raw.githubusercontent.com/lynxthecat/cake-autorate/master/setup.sh
Downloading 'https://raw.githubusercontent.com/lynxthecat/cake-autorate/master/setup.sh'
Connecting to 185.199.111.133:443
Writing to 'setup.sh'
setup.sh             100% |*******************************|  6806   0:00:00 ETA
Download completed (6806 bytes)
root@OpenWrt-1:~/cake-autorate# ./setup.sh
-ash: ./setup.sh: Permission denied
root@OpenWrt-1:~/cake-autorate# chmod +x setup.sh
root@OpenWrt-1:~/cake-autorate# ./setup.sh
Detected Operating System: openwrt
Installation directories for detected Operating System:
  - Script prefix: /root/cake-autorate
  - Config prefix: /root/cake-autorate
Continue with installation? [Y/n]
Installing cake-autorate using /root/cake-autorate (script) and /root/cake-autorate (config) directories...
Previous configuration present - keep it? [Y/n]
Using prior configuration

3.2.0-PRERELEASE successfully installed, but not yet running

Start the software manually with:
   cd /root/cake-autorate; ./cake-autorate.sh
Run as a service with:
   service cake-autorate enable; service cake-autorate start

root@OpenWrt-1:~/cake-autorate# service cake-autorate stop
Command failed: Not found
root@OpenWrt-1:~/cake-autorate# service cake-autorate start
```

### PR DESCRIPTION
**Pull request by rany2**
Fixes #293

### CODE REVIEW

**Comment by lynxthecat** [comment]
I tested and it seems to work.

**Comment by rany2** [comment]
@lynxthecat that's good, I didn't test because I don't run cake-autorate on asuswrt/openwrt; so i needed your review. That's good, I just want to make some minor changes

File: setup.sh [hunk 1]
```diff
@@ -118,10 +123,35 @@ main() {
...
+		printf "Local setup.sh differs from the one in the repository\n"
+		if grep -q ' CAKE_AUTORATE_SETUP_SH_SUPPORT_SELF_REPLACING$' "${tmp}/setup.sh"
+		then
+			printf "Self-replacing setup.sh and restarting while preserving the current environment...\n"
+			trap - EXIT INT TERM
+			__CAKE_AUTORATE_SETUP_SH_EXEC_TMP="${tmp}" \
+			__CAKE_AUTORATE_SETUP_SH_EXEC_COMMIT="${commit}" \
+				exec "${SHELL}" "${tmp}/setup.sh" "${REPOSITORY}" "${BRANCH}"
+		else
+			printf "Self-replacing not fully supported, restarting with the new setup.sh...\n"
```
Comment (lynxthecat) [inline setup.sh:148]:
```suggestion
			printf "Self-replacing not fully supported. Restarting with the new setup.sh...\n"
```

**Review by lynxthecat — APPROVED** [review]

**Comment by lynxthecat** [comment]
Intimidating. Please feel free to merge!

**Comment by rany2** [comment]
BTW there are mistakes in this like how $SHELL is used, it points to the default shell not the current shell; I could get the current shell from /proc/$$/cmdline but that's not great

**Comment by rany2** [comment]
There's no need to use the current shell in the first place, I think it runs fine with /bin/sh on asuswrt now.

**Pull request merged**